### PR TITLE
nixos/jitsi-meet: init

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
**Please Read: The previous PR exploded with feature requests and nitpicks. The goal of this PR is to provide a minimum viable product that can be merged into nixpkgs. Additional features and improvements can be added as followup PRs later.**

A few friends and me are using the Jitsi Meet module by mmilata since a while and I want to get it merged into nixpkgs.
However I don't have the resources to maintain Jitsi Meet in nixpkgs long-term. I am still looking for someone to help me with that.

### TODO

- [ ] Update packages
- [ ] Let's Encrypt based shared certificates for prosody and the ji* services
- [ ] Manual section / documentation on how to set up

### Nice to have
- [ ] One more extensive test would be good, at least with two clients joining the room

### Out of scope for this PR

- Build packages from source
- Reliable OCR-based tests for video

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/jicofo.nix
Comment (Ma27) [inline nixos/modules/services/networking/jicofo.nix:116]: I usually dislike nitpicks (you don't have to fix it), but I'd like to point out that `ExecStart=` may be better here (since this is a single command), it's still possible to use a multiline-string for it.

Using `serviceConfig.ExecStart` has the benefit that you directly see what's happening in the service when running `systemctl cat jicofo` rather than having to `cat` a copied store-path.

**Review by Ma27 — COMMENTED** [review]
Thanks a lot for taking care of this!

I just tested it in a local network and it appears to work fine when enabling TLS. I'm planning to self-host a jitsi instance soon, I can probably provide even more feedback (if I get to this before this PR gets merged).

Below are some comments/thoughts I had while reviewing your changes. Since the module works fine overall, I don't consider those mandatory to get this PR merged though.

File: pkgs/servers/jitsi-videobridge/default.nix
Comment (petabyteboy) [reply pkgs/servers/jitsi-videobridge/default.nix:44]: Yes. As described in the initial comment, I don't have the resources to maintain this on my own. The original author also removed themselves from the maintainers list before closing the old PR. 
If someone is willing to take the main maintainership, I will join in a supportive role (testing minor updates, fixing bugs, etc.). 
I want to prevent a situation where this gets merged with me as the only maintainer and it rots away in nixpkgs.

Comment (Ma27) [reply pkgs/servers/jitsi-videobridge/default.nix:44]: Thanks for the clarification. I guess I missed that originally, sorry!

File: nixos/tests/jitsi-meet.nix
Comment (petabyteboy) [reply nixos/tests/jitsi-meet.nix:22]: I think there is no universal standard among NixOS modules here, and one could probably have a long and tiring debate. Other modules for services that can not be used without TLS either don't set it in the module. We have a note in the documentation that users _probably_ want to enable it.
Some `rg enableACME nixos` reveals that most modules that _do_ set it, just set it to `mkDefault true`, which I will do now as well.

File: nixos/modules/services/networking/jicofo.nix
Comment (petabyteboy) [reply nixos/modules/services/networking/jicofo.nix:146]: Done, thanks

File: nixos/modules/services/networking/jicofo.nix
Comment (petabyteboy) [reply nixos/modules/services/networking/jicofo.nix]: Done.

File: nixos/modules/services/web-apps/jitsi-meet.nix
Comment (petabyteboy) [reply nixos/modules/services/web-apps/jitsi-meet.nix:313]: Not sure. The user would have to make quite a lot of changes to make it work with certificate verification, because the videobridge connects to localhost I think. If the user is able to change all that, they can just as well use mkForce here to change the option. One could say that it is better this way because the user is warned that this option really shouldn't be changed, because it will most probably break the service.
So I have a slight preference for leaving it as it is.
But I'm looking into the possibility of using a proper domain name and certificate to connect to the server, making i unnecessary to set the option here.

Comment (Ma27) [reply nixos/modules/services/web-apps/jitsi-meet.nix:313]: Fair enough then. I assumed that this was an opinionated default (that should be easily changeable), but apparently it isn't :)

**Comment by Ma27** [comment]
The change seems fine now, thanks!

I'd prefer to get a bit more feedback from other community members, other than that this patch seems fine.

File: nixos/modules/services/networking/jicofo.nix
Comment (puckipedia) [inline nixos/modules/services/networking/jicofo.nix:113]: This literally passes `$(cat /var/lib/jitsi-meet/jicofo-component-secret)` in, somehow i suspect that's not meant to happen :p

Comment (petabyteboy) [reply nixos/modules/services/networking/jicofo.nix:113]: Can you explain a bit why you think this should not happen?

Comment (puckipedia) [reply nixos/modules/services/networking/jicofo.nix:113]: As in, my `ps aux` showed that `jicofo` got the above \*string\* (not the contents of the file) as secret, which broke jicofo connecting to the local prosody.

Comment (petabyteboy) [reply nixos/modules/services/networking/jicofo.nix:113]: Oh, I think this was using the script attribute before, so this would have been run in a shell. I'll revert to using the script attribute again, thanks for finding this!

Comment (Ma27) [reply nixos/modules/services/networking/jicofo.nix]: Oh sorry, my mistake for not realizing this problem during my review!

**Comment by puckipedia** [comment]
One more issue I noticed: the videobridge dynamically tries to load a .so, which binds to openssl, which can't be located. This breaks crypto acceleration somewhat..

**Comment by sbourdeauducq** [comment]
> This breaks crypto acceleration somewhat..

Could we merge this and then fix small issues like that later?

**Comment by ryantm** [comment]
I just started using this set of patches on top of master and it works great for me.

**Review by Ma27 — APPROVED** [review]
@petabyteboy I'm in favor of merging this. Is there anything missing in your opinion or do you agree? :)

**Comment by petabyteboy** [comment]
> @petabyteboy I'm in favor of merging this. Is there anything missing in your opinion or do you agree? :)

We don't have any maintainers. Otherwise it would be okay if you merged it, but I would want to complete the todos in the description before I would feel comfortable merging it.

**Comment by dpausp** [comment]
I'm running some Jitsi instances for my company https://flyingcircus.io and our customers. Last week, I switched our test system from a docker-based setup to the code from this PR with updated package versions and it's working well so far. I'd like to see this PR merged. After that, I can take over as a maintainer and work on additional features.

**Review by ryantm — APPROVED** [review]

**Pull request merged**
